__tests__/test-utils/teardown-e2e.js

```diff
@@ -0,0 +1,294 @@
+          os: 'default',
+          uri: `https://${LOGS_S3_BUCKET}.s3.amazonaws.com/${uploadedLogsFilename}`
+        }
+      ]
+    })
+  }
+
+  let fetchResponse
+  try {
+    fetchResponse = await fetch(
+      process.env.MS_TEAMS_WEBHOOK_URL,
+      {
+        method: 'POST',
+        body: JSON.stringify({
+          '@context': 'https://schema.org/extensions',
+          '@type': 'MessageCard',
+          themeColor: '0072C6',
+          title: `${repo} e2e test ${testResults.success ? 'passed. ✅' : 'failed. ❌'}`,
+          text: `${testResults.numPassedTests} / ${testResults.numTotalTests} tests passed`,
+          potentialAction: actions
+        })
+      }
+    )
+  } catch (e) {
+    console.error('Failed to post to MS Teams: ', e)
+    throw e
+  }
+
+  if (fetchResponse.status >= 400) {
+    console.error('Received an error code for MS Teams post!')
+    throw new Error('Post to MS Teams failed!')
+  }
+
+  console.log('successfully posted to MS Teams')
+}
+
+/**
+ * Upload the zip file containing the lgos to a slack channel.
+ */
+async function uploadToSlack () {
+  if (!slackConfigured) return
+
+  const slack = new Slack({ token: process.env.SLACK_TOKEN })
+
+  console.log('uploading log zipfiles to slack')
+  try {
+    await slack.files.upload({
+      channels: process.env.SLACK_CHANNEL,
+      file: fs.createReadStream(logsZipfile),
+      filename: uploadedLogsFilename,
+      filetype: 'zip',
+      initial_comment: `View build logs here: ${process.env.TRAVIS_BUILD_WEB_URL}`
+    })
+  } catch (e) {
+    console.error('failed to upload logs to slack!')
+    throw e
+  }
+
+  console.log('successfully uploaded logs to slack')
+}
+
+/**
+ * Find all outputted log files, zip them up and then upload them to a defined
+ * slack or MS Teams channel (if defined)
+ */
+function uploadLogs () {
+  if (!(slackConfigured || msTeamsConfigured)) {
+    console.warn('Log upload environment variables undefined, not uploading logs anywhere!')
+    return
+  }
+
+  const output = fs.createWriteStream(logsZipfile)
+  const archive = archiver('zip')
+
+  // handle errors if they happen
+  archive.on(
+    'error',
+    makeUploadFailureHandler('An error occurred while archiving')
+  )
+  output.on(
+    'error',
+    makeUploadFailureHandler('An error occurred while writing the zip file')
+  )
+
+  output.on('close', () => {
+    console.log(`Successfully zipped ${archive.pointer()} total bytes`)
+    Promise.all([uploadToSlack(), uploadToMicrosoftTeams()])
+      .then(() => {
+        console.log('successfully uploaded logs')
+      })
+      .catch(err => {
+        if (err) {
+          return makeUploadFailureHandler(
+            'An error occurred while uploading the logs'
+          )(err)
+        }
+      })
+  })
+
+  archive.pipe(output)
+
+  // glob all files in test folder and put them in root folder of zip file
+  archive.directory(getTestFolderFilename(), false)
+  archive.finalize()
+}
+
+module.exports = function () {
```

Comment:
Same comment as for setup file. Is it a node convention to put module.exports at the end of the file?

Reply:
In certain cases it'd be fine to have it at the top, but I think in general it's a good idea to have it at the bottom because not everything gets hoisted. See https://stackoverflow.com/questions/23791749/why-must-a-modules-exports-be-declared-at-the-bottom-of-a-file and https://scotch.io/tutorials/understanding-hoisting-in-javascript. I'll add comments about this exported function tho.